frontend/include/chpl/framework/ErrorWriter.h

```diff
@@ -330,16 +350,37 @@ class ErrorWriterBase {
     This function accepts any type for which location can be inferred,
     for both the main location and the highlights.
    */
-  template <typename LocPlace, typename LocHighlight = const uast::AstNode*>
+  template <typename LocHighlight = const uast::AstNode*, typename LocPlace>
   void code(const LocPlace& place,
-                const std::vector<LocHighlight>& toHighlight = {}) {
+            const std::vector<LocHighlight>& toHighlight = {}) {
     std::vector<Location> ids(toHighlight.size());
     std::transform(toHighlight.cbegin(), toHighlight.cend(), ids.begin(), [&](auto node) {
       return errordetail::locate(context, node);
     });
     writeCode(errordetail::locate(context, place), ids);
   }
 
+  /**
+    Specialziation of ErrorWriter::code for printing definitions. Since
+    definitions may be fairly length (e.g., a variable with a multi-line
+    initializer), this will only print the first line of the definition.
+   */
+  template <typename LocPlace>
+  void codeForDef(const LocPlace& place) {
+    code<LocPlace>(justOneLine(place), { place });
+  }
+
+  /**
+    Specialziation of ErrorWriter::code for "bla bla defined here".
+    Doesn't underline, but also doesn't print the whole thing. This
+    is important, for instance, in the case that a record declaration
+    is being printed. We don't want to dump all the fields / methods
+    as part of the errror.
```

Comment:
One thing I am confused by here is that I thought that printing the code for a function declaration would print only the signature. Is that accurate? If so, wouldn't we want the same approach on records / classes?

Reply:
I am at the mercy of the parser and its location reporting mechanisms. Seems like right now, a record’s “location” is the whole record, but a “function’s” location is just the signature / declaraton / part before {}.  Code printing an AST node will print only the function signature, but a whole record. To be cautious and to avoid huge outputs, I always print only one line.